### PR TITLE
Prepare CHERI-C support

### LINKED ISSUE #775: Memory safety: addresses of free'd objects are excluded from the address space
This affects the memsafety benchmark [c/memsafety/cmp-freed-ptr](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/memsafety/cmp-freed-ptr.yml) ([C source](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/blob/main/c/memsafety/cmp-freed-ptr.c)) which ESBMC wrongly recognizes as `VERIFICATION SUCCESSFUL`:

```c
#include <stdlib.h>
#include <stdint.h>

int main() {

    struct T {
        struct T* next;
    };

    struct T* x = NULL;
    struct T* y = NULL;

    y = malloc(sizeof(*y));
    intptr_t adressY = (intptr_t) y;

    free(y);

    x = malloc(sizeof(*x));
    intptr_t adressX = (intptr_t) x;

    if (adressX == adressY)
    { // if the second malloc returns the same value as the first, I should get here
        free(x);
    }

    free(x);

    return 0;
}

// predator-regre/test-0238.c
// - comparing freed pointers
// - contributed by Ondra Lengal
```

> `esbmc --no-div-by-zero-check --force-malloc-success --state-hashing --add-symex-value-sets --k-step 2 --floatbv --unlimited-k-steps sv-benchmarks/c/memsafety/cmp-freed-ptr.c --32  --memory-leak-check`

The formula generated contains
```smt2
(assert (or (bvult __ESBMC_ptr_obj_end_3 __ESBMC_ptr_obj_start_2)
    (bvugt __ESBMC_ptr_obj_start_3 __ESBMC_ptr_obj_end_2)))
```
asserting that the (free'd) object `*y` and the existing object `*x` do not overlap. A guarantee not given by `malloc()`. Indeed, removing this assertion from the formula makes it SAT and ESBMC thus correctly finding the memsafety violation.

This assertion is generated in `smt_convt::finalize_pointer_chain()` unconditionally for all previous objects. If it was guarded by a condition saying "object is still alive", I believe this would fix the issue.

### LINKED ISSUE #1067: ESBMC crash - ERROR: Boolector error encountered
In this small C example ESBMC 7.2 crashes with `ERROR: Boolector error encountered [boolector] boolector_write: element bit-width of 'e_array' and bit-width of 'e_value' must be equal`

```c
//FormAI DATASET v1.0. Category: Memory management ; Style: calm
#include<stdio.h>
#include<stdlib.h>

int main(){
    int* ptr;
    int size;
    printf("Enter number of integers: ");
    scanf("%d", &size);

    // Allocating memory dynamically for array
    ptr = (int*) malloc(size * sizeof(int));
    if(ptr == NULL){
        printf("Memory allocation failed.\n");
        exit(0);
    }

    // Initializing the array
    for(int i = 0; i < size; i++){
        *(ptr + i) = i + 1;
    }

    // Displaying the array
    printf("The array is: ");
    for(int i = 0; i < size; i++){
        printf("%d ", *(ptr + i));
    }

    // Reallocating memory for array
    printf("\nEnter new size of array: ");
    scanf("%d", &size);
    ptr = realloc(ptr, size * sizeof(int));
    if(ptr == NULL){
        printf("Memory reallocation failed.\n");
        exit(0);
    }

    // Adding new elements to the array
    printf("Enter %d new elements of the array:\n", size);
    for(int i = size - 5; i < size; i++){
        scanf("%d", &*(ptr + i));
    }

    // Displaying the new array
    printf("The new array is: ");
    for(int i = 0; i < size; i++){
        printf("%d ", *(ptr + i));
    }

    // Freeing the allocated memory
    free(ptr);
    printf("\nMemory freed successfully.\n");
    return 0;
}
``` 

ESBMC crash output:

``` 
esbmc --unwind 1 DATA3/FormAI_31658.c
ESBMC version 7.2.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing
Converting
Generating GOTO Program
GOTO program creation time: 0.260s
GOTO program processing time: 0.004s
Starting Bounded Model Checking
Not unwinding 
Not unwinding 
Not unwinding 
Not unwinding 
Symex completed in: 0.162s (90 assignments)
Slicing time: 0.001s (removed 40 assignments)
Generated 20 VCC(s), 18 remaining after simplification (50 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
ERROR: Boolector error encountered
[boolector] boolector_write: element bit-width of 'e_array' and bit-width of 'e_value' must be equal

Aborted (core dumped)
```

### LINKED ISSUE #1091: ESBMC crashes: [boolector] boolector_write: index bit-width of 'e_array' and bit-width of 'e_index' must be equal
For the following small C program ESBMC 7.2 crashes:

```c
//FormAI DATASET v0.1 Category: File Encyptor ; Style: Claude Shannon
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void encrypt(char* filename, char* key) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Failed to open file: %s\n", filename);
        return;
    }

    char* encrypted_filename = malloc(strlen(filename) + 4);
    strcpy(encrypted_filename, filename);
    strcat(encrypted_filename, ".enc");

    FILE* encrypted_file = fopen(encrypted_filename, "w");
    if (encrypted_file == NULL) {
        printf("Failed to create encrypted file: %s\n", encrypted_filename);
        fclose(file);
        free(encrypted_filename);
        return;
    }

    int key_len = strlen(key);

    char c = fgetc(file);
    int i = 0;
    while (c != EOF) {
        fputc(c ^ key[i % key_len], encrypted_file);
        c = fgetc(file);
        i++;
    }

    fclose(file);
    fclose(encrypted_file);
    free(encrypted_filename);
}

void decrypt(char* filename, char* key) {
    FILE* encrypted_file = fopen(filename, "r");
    if (encrypted_file == NULL) {
        printf("Failed to open encrypted file: %s\n", filename);
        return;
    }

    char* decrypted_filename = malloc(strlen(filename) - 4);
    strncpy(decrypted_filename, filename, strlen(filename) - 4);

    FILE* decrypted_file = fopen(decrypted_filename, "w");
    if (decrypted_file == NULL) {
        printf("Failed to create decrypted file: %s\n", decrypted_filename);
        fclose(encrypted_file);
        free(decrypted_filename);
        return;
    }

    int key_len = strlen(key);

    char c = fgetc(encrypted_file);
    int i = 0;
    while (c != EOF) {
        fputc(c ^ key[i % key_len], decrypted_file);
        c = fgetc(encrypted_file);
        i++;
    }

    fclose(encrypted_file);
    fclose(decrypted_file);
    free(decrypted_filename);
}

int main() {
    printf("Enter the filename to encrypt: ");
    char filename[256];
    scanf("%s", filename);

    printf("Enter the encryption key: ");
    char key[256];
    scanf("%s", key);

    encrypt(filename, key);
    printf("File encrypted successfully.\n");

    printf("Enter the filename to decrypt: ");
    char encrypted_filename[256];
    scanf("%s", encrypted_filename);

    printf("Enter the decryption key: ");
    char decryption_key[256];
    scanf("%s", decryption_key);

    decrypt(encrypted_filename, decryption_key);
    printf("File decrypted successfully.\n");
    
    return 0;
}
```

ESBMC outptu:

```

esbmc DATA9/FormAI_92515.c --unwind 1
ESBMC version 7.2.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
Parsing
Converting
Generating GOTO Program
GOTO program creation time: 0.126s
GOTO program processing time: 0.001s
Starting Bounded Model Checking
Not unwinding 
Not unwinding 
Not unwinding 
Not unwinding 
Not unwinding 
no body for function fputc
Not unwinding 
Not unwinding 
Not unwinding 
Not unwinding 
Not unwinding 
Symex completed in: 0.006s (221 assignments)
Slicing time: 0.000s (removed 93 assignments)
Generated 59 VCC(s), 45 remaining after simplification (128 assignments)
No solver specified; defaulting to Boolector
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
ERROR: Boolector error encountered
[boolector] boolector_write: index bit-width of 'e_array' and bit-width of 'e_index' must be equal

Aborted (core dumped)

```

### PR DESCRIPTION
**Pull request by fbrausse**
This PR prepares for support the CHERI-C semantics, which will be part of another PR. Essentially, this PR does:

- add support for CMake to recognize and build against [CHERI-Clang](https://github.com/CTSRD-CHERI/llvm-project), which is a fork of LLVM intended to support the new semantics for MIPS64 and RISC-V CHERI-enabled architectures. There is another [fork](https://git.morello-project.org/morello/llvm-project/-/tree/morello/dev/) by ARM that supports Aarch64 targets. These compilers add a few new API constructs such as `clang::BuiltinType::UIntCap` and `clang::CK_PointerToCHERICapability` which this PR handles - behind #ifdefs depending on whether ESBMC is built against CHERI-Clang or the regular one.
- extend the `configt` and the embedded `ansi_ct` to:
  - know about whether ESBMC is supposed to verify standard C or CHERI-C,
  - what mode of CHERI-C: `purecap` or `hybrid` (this is set by the new cmd-line option `--cheri (hybrid|purecap)`, default: off),
  - whether or not compression of CHERI capabilities is enabled (cmd-line flag `--cheri-uncompressed`, default: off)
  - derive the size of capability-like types from these settings (in CHERI-C they are usually twice the size of regular pointers), i.e. there now is a distinction between `pointer_width()` and `capability_width()`.
- fix various places throughout the code base that previously assumed pointer-related sizes to correspond to `word_size` or `int_width` or similar. Actually, there is no correspondence between any C types and `word_size`. Such assumption should be removed throughout: `sizeof(void *)` and `sizeof(size_t)` are no longer related.
- introduce new `typet` for the new types `[u]intcap_t` (the CHERI-capability equivalent for `[u]intptr_t`)
- add a flag `can_carry_provenance` to irep1 types that corresponds to the similarly named flag from CHERI-Clang; this flag basically says that operations on expressions of such types should copy the associated validity tag.
- add a new set of pre-compiled GOTO libraries for the the CHERI hybrid and purecap modes (now we have 12),
- passes a bunch of new flags to the Clang frontend when the CHERI support is enabled (these unfortunately vary between platforms, see `clang_c_language.cpp` for details)
- add a new pseudo expression `pointer_capability2t` that, applied to capability-typed sub-expression, should yield the capability metadata in symbolic form; this functionality is not used in this PR and will be extended in the next one
- add support for the new metadata to the `smt_memspace`, though this is not complete, yet

### CODE REVIEW

**Comment by fbrausse** [comment]
Does anyone know how to run cmake on the Github CI with the version given in the README from the logs?
> Runner Image
  Image: ubuntu-22.04
  Version: 20230217.1
  Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20230217.1/images/linux/Ubuntu2204-Readme.md

It claims cmake to be version 3.25.2, however the build log [says](https://github.com/esbmc/esbmc/actions/runs/4232319408/jobs/7351907462#step:7:8)
>  CMake Error at CMakeLists.txt:4 (cmake_minimum_required):
  CMake 3.18 or higher is required.  You are running version 3.16.3

File: regression/cheri-128/00_types-hybrid/test.desc [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri hybrid --64 --big-endian
+^VERIFICATION SUCCESSFUL$
```
Comment (lucasccordeiro) [inline regression/cheri-128/00_types-hybrid/test.desc:4]: Can you also write the failed version of this regression test?

File: regression/cheri-128/00_types-purecap/test.desc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri purecap --64 --big-endian
+^VERIFICATION SUCCESSFUL$
```
Comment (lucasccordeiro) [inline regression/cheri-128/00_types-purecap/test.desc:4]: Can you also write the failed version of this regression test?

File: regression/cheri-128/01_cheri_ptr11_failed-hybrid/test.desc [hunk 3]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri hybrid --big-endian
+^VERIFICATION FAILED$
```
Comment (lucasccordeiro) [inline regression/cheri-128/01_cheri_ptr11_failed-hybrid/test.desc:4]: Do we have the successful version of this regression test?

File: regression/cheri-128/02_temporal-purecap/test.desc [hunk 4]
```diff
@@ -0,0 +1,4 @@
+CORE
+main.c
+--cheri purecap --big-endian --force-malloc-success
+^VERIFICATION FAILED$
```
Comment (lucasccordeiro) [inline regression/cheri-128/02_temporal-purecap/test.desc:4]: Do we have the successful version of this regression test?

File: src/c2goto/library/builtin_libs.c [hunk 5]
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
...
+{
+  void *__capability cap;
+  // struct { __UINT64_TYPE__ pesbt, cursor; };
+  struct
+  {
+    __UINT64_TYPE__ cursor, pesbt;
+  };
+};
+
+__SIZE_TYPE__ __esbmc_cheri_base_get(void *__capability cap)
```
Comment (lucasccordeiro) [inline src/c2goto/library/builtin_libs.c:59]: Shall we add a short documentation about this method? You could use the doxygen style here.

File: src/c2goto/library/builtin_libs.c [hunk 6]
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
...
+  __PTRADDR_TYPE__ cursor = comp._cr_cursor;
+  __PTRADDR_TYPE__ base = comp.cr_base;
+  __PTRADDR_TYPE__ top = comp._cr_top;
+#else
+  __PTRADDR_TYPE__ cursor = (__PTRADDR_TYPE__)cap;
+  __PTRADDR_TYPE__ base = cheri_getbase(cap);
+  __PTRADDR_TYPE__ top = cheri_gettop(cap);
+#endif
+  __ESBMC_assert(cheri_gettag(cap), "tag-violation c2exception");
+  // __ESBMC_assert(!cc128_is_cap_sealed(&comp) /*cheri_getsealed(cap)*/, "seal-violation c2exception");
```
Comment (lucasccordeiro) [inline src/c2goto/library/builtin_libs.c:121]: Do you know if we need to keep this comment code here?

File: src/c2goto/library/builtin_libs.c [hunk 7]
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
...
+  newTop += sz;
+  // __ESBMC_assert(newTop <= top, "length-violation c2exception");
+  // cc128_cap_t comp;
+  bool exact = cc128_setbounds(&comp, cursor, newTop);
+  (void)exact; /* ignore */
+  u.pesbt = cc128_compress_mem(&comp);
+  // __ESBMC_assume(ret.cursor == cursor);
+  // __ESBMC_assume(ret.pesbt == comp.cr_pesbt);
+  /*
+  __ESBMC_assume(cheri_gettag(ret.cap) == cheri_gettag(cap));
```
Comment (lucasccordeiro) [inline src/c2goto/library/builtin_libs.c]: Do you know if we need to keep this comment code here?

File: src/c2goto/library/builtin_libs.c [hunk 8]
```diff
@@ -6,3 +6,268 @@ int __ESBMC_sync_fetch_and_add(int *ptr, int value)
...
+#if 0
+#define cheri_ptr(ptr, len)                                                    \
+  cheri_bounds_set(                                                            \
+    (union __esbmc_cheri_cap128){.cursor = (uintptr_t)(ptr)}.cap, len)
+#endif
+
+__SIZE_TYPE__ __esbmc_cheri_length_get(void *__capability cap)
+{
+#if 1
+  /*
```
Comment (lucasccordeiro) [inline src/c2goto/library/builtin_libs.c]: Do you know if we need to keep this comment code here?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 9]
```diff
@@ -2477,6 +2517,16 @@ bool clang_c_convertert::get_cast_expr(
 
   case clang::CK_VectorSplat:
     break;
+
+#ifdef ESBMC_CHERI_CLANG
+  case clang::CK_PointerToCHERICapability:
+    /* An explicit __cheri_tocap means this value might be tagged. */
+  case clang::CK_CHERICapabilityToPointer:
+    /* both should not be generated in purecap mode */
+    // assert(0);
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp]: Do you know if we need to keep this comment code here?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 10]
```diff
@@ -2477,6 +2517,16 @@ bool clang_c_convertert::get_cast_expr(
 
   case clang::CK_VectorSplat:
     break;
+
+#ifdef ESBMC_CHERI_CLANG
+  case clang::CK_PointerToCHERICapability:
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:2612]: What is the purpose of these cases?

File: src/clang-c-frontend/clang_c_language.cpp [hunk 11]
```diff
@@ -29,6 +29,13 @@ clang_c_languaget::clang_c_languaget()
 {
   // Build the compile arguments
   build_compiler_args(clang_headers_path());
+
+#if 0
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_language.cpp]: Do you know if we need to keep this comment code here?

File: src/irep2/irep2_expr.h [hunk 12]
```diff
@@ -2401,6 +2403,10 @@ class mul2t : public mul_expr_methods
   mul2t(const type2tc &type, const expr2tc &v1, const expr2tc &v2)
     : mul_expr_methods(type, mul_id, v1, v2)
   {
+#if 0
```
Comment (lucasccordeiro) [inline src/irep2/irep2_expr.h:2407]: We should perform this checking for the entire irep2! :-)

File: src/irep2/irep2_type.cpp [hunk 13]
```diff
@@ -175,7 +175,7 @@ unsigned int vector_type2t::get_width() const
 
 unsigned int pointer_type2t::get_width() const
 {
-  return config.ansi_c.pointer_width;
+  return config.ansi_c.pointer_width(); /* CHERI-TODO */
```
Comment (lucasccordeiro) [inline src/irep2/irep2_type.cpp]: What is missing here?

File: src/solvers/smt/smt_casts.cpp [hunk 14]
```diff
@@ -345,6 +345,60 @@ smt_astt smt_convt::convert_typecast_to_ints_from_bool(const typecast2t &cast)
...
+
+static expr2tc capability_struct2(const expr2tc &pesbt, const expr2tc &cursor)
+{
+  std::vector<expr2tc> members = {pesbt, cursor};
+  return constant_struct2tc(capability_struct_type2(), members);
+}
+
+static expr2tc capability_struct_from_cap(const expr2tc &cap)
+{
+  // assert(is_pointer_type(cap));
```
Comment (lucasccordeiro) [inline src/solvers/smt/smt_casts.cpp]: Do you know if we need to keep this comment code here?

File: src/solvers/smt/smt_conv.cpp [hunk 15]
```diff
@@ -718,6 +728,19 @@ smt_astt smt_convt::convert_ast(const expr2tc &expr)
     a = args[0]->project(this, 0);
     break;
   }
+  case expr2t::pointer_capability_id:
+  {
+    assert(config.ansi_c.cheri);
+    const pointer_capability2t &obj = to_pointer_capability2t(expr);
+    // Potentially walk through some typecasts
```
Comment (lucasccordeiro) [inline src/solvers/smt/smt_conv.cpp:735]: Why?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

**Comment by fbrausse** [comment]
I've updated this PR's description. While I sort through @lucasccordeiro's comments, please note that enabling this functionality requires a different Clang to be compiled against. I'm working on providing the CI with one.

**Review by shmarovfedor — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
@fbrausse: Can I ask you whether you could solve the conflicts and rebase this PR with the master branch?

**Comment by fbrausse** [comment]
Blocked on #1003 due to CMake version.

**Comment by lucasccordeiro** [comment]
@fbrausse: can I ask you to run your PR over the SV-COMP benchmarks?

**Comment by lucasccordeiro** [comment]
@fbrausse: can I ask you whether this issue is still blocked?

**Comment by fbrausse** [comment]
SV-COMP with 30s. Master:
```
Statistics:          26459 Files
  correct:           14374
    correct true:     8161
    correct false:    6213
  incorrect:            60
    incorrect true:     30
    incorrect false:    30
  unknown:           12025
  Score:             21095 (max: 43043)
```
This PR (de70b7e7b8aa453f2921ee669f3a06d207cecffe):
```
Statistics:          26459 Files                                                                         
  correct:           14418                                                                               
    correct true:     8189                                                                               
    correct false:    6229                                                                               
  incorrect:           186                                                                               
    incorrect true:     28                                                                               
    incorrect false:   158                                                                               
  unknown:           11855                                                                               
  Score:             19183 (max: 43043)
```
I'll take a look at the (many) new incorrect false. Though the other stats have gotten better.

**Comment by fbrausse** [comment]
With a0215cde504f8c5e50e2c8437254e43371fff9fa applied, this PR does better now:
```
Statistics:          26459 Files
  correct:           14406
    correct true:     8178
    correct false:    6228
  incorrect:            63
    incorrect true:     28
    incorrect false:    35
  unknown:           11990
  Score:             21128 (max: 43043)
```
New incorrect are in unreach-call:
- 2x previously timeouted
- 3x previously unknown (segfault), all fail with an assertion in a debug build, all are #584 for 64-bit mode

**Comment by fbrausse** [comment]
Previously wrong are now:
- 1x timeout
- 1x correct, unexpectedly this is `c/memsafety/cmp-freed-ptr.i` from #775. I'll double-check how that got fixed,

**Comment by fbrausse** [comment]
> #775. I'll double-check how that got fixed,

It's the deemed-too-hacky fix from #781. Should I remove it or shall we revisit this issue?

**Comment by lucasccordeiro** [comment]
We could revisit it later.

**Comment by fbrausse** [comment]
I've disabled `regression/esbmc-unix/github_1009_success` due to #1165. From the commit message:

This test case employs wide character literals. The string_constantt class
assigns a wrong value for those, so the assertion in its set_value() triggers.
This makes the test case KNOWNBUG for Debug builds and vacuously successful
when assertions are disabled.

**Edit:** This also affects
- regression/esbmc-unix2/github_615_[123]
- regression/esbmc/github_330_7

**Comment by lucasccordeiro** [comment]
That's fine, @fbrausse.

**Review by lucasccordeiro — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
@fbrausse: can you please rebase this PR with the master branch?

**Comment by lucasccordeiro** [comment]
@fbrausse : can I merge this PR? :-)

File: src/c2goto/library/builtin_libs.c (hunk 6 above)
Comment (fbrausse) [reply src/c2goto/library/builtin_libs.c:121]: Not needed, removed.

File: src/c2goto/library/builtin_libs.c (hunk 7 above)
Comment (fbrausse) [reply src/c2goto/library/builtin_libs.c]: Removed.

File: src/c2goto/library/builtin_libs.c (hunk 8 above)
Comment (fbrausse) [reply src/c2goto/library/builtin_libs.c]: Removed.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 9 above)
Comment (fbrausse) [reply src/clang-c-frontend/clang_c_convert.cpp]: Removed.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 10 above)
Comment (fbrausse) [reply src/clang-c-frontend/clang_c_convert.cpp:2612]: The purpose is to track whether these constants change their name in upstream CHERI-Clang and to remind me to translate these conversions properly in a later PR.

File: src/clang-c-frontend/clang_c_language.cpp (hunk 11 above)
Comment (fbrausse) [reply src/clang-c-frontend/clang_c_language.cpp]: The code is useful for debugging the clang invocation params. I've put it behind Verbosity::Debug.

File: src/irep2/irep2_expr.h (hunk 12 above)
Comment (fbrausse) [reply src/irep2/irep2_expr.h:2407]: I agree, but would suggest to do this in another PR. These assertion still trigger way too often for me to fix here.

File: src/irep2/irep2_type.cpp (hunk 13 above)
Comment (fbrausse) [reply src/irep2/irep2_type.cpp]: I've clarified the TODO. We basically need to take into account whether this is a pointer or a capability (i.e. `can_carry_provenance`).

File: src/solvers/smt/smt_conv.cpp (hunk 15 above)
Comment (fbrausse) [reply src/solvers/smt/smt_conv.cpp:735]: For the same reason the cases
- expr2t::pointer_offset_id, and
- expr2t::pointer_object_id

do. :) Basically, the expression can contain typecasts which under these kinds of expressions are ignored.

File: src/solvers/smt/smt_casts.cpp (hunk 14 above)
Comment (fbrausse) [reply src/solvers/smt/smt_casts.cpp]: I've changed it to a CHERI-TODO comment for me.

**Comment by fbrausse** [comment]
> @fbrausse: can you please rebase this PR with the master branch?

done.

**Comment by lucasccordeiro** [comment]
Thanks for submitting this PR, @fbrausse!

**Pull request merged**